Allow integer and bool values for filtering images in ec2_ami_info

jainnikhil30 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
CC @san7ket @Akasurde 
Fixes: https://github.com/ansible-collections/amazon.aws/issues/28
This is taken from the PR https://github.com/ansible/ansible/pull/48654 which was created in the ansible repo before the collection move. I have added the integration tests as suggested by @jillr 

Also the change suggested by @mrjbanksy was not needed for the in-built bool variables. I think in case of tags it still goes as strings ? I have a test case where it looks for is-public filter which is a bool and it works with `str(v).lower()`. Infact if I use `str(v).capatalize()`, the inbuilt boolean variables don't work for filtering.
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ec2_ami_info

Akasurde commented on changelogs/fragments/28-ec2_ami_info_int_bool_filter.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - ec2_ami_info - allow integer and bool values for filtering images
```
```suggestion
  - ec2_ami_info - allow integer and bool values for filtering images (https://github.com/ansible/ansible/issues/43570).
```

briantist commented:
This is great, happy to see it! As I mentioned in #88 , this should also affect `ec2_instance` and `ec2_instance_info`, perhaps other modules I'm not aware of.

It might make sense to update their tests as well; if nothing else the changelog should reflect the update for those modules too.

jainnikhil30 commented:
@tremble @jillr Could you please take a look at this ?

tremble commented:
Looking specifically at the comment @mrjbanksy made on the original PR: "It's complicated".

What's I think has happened is that when `SomeKey: true` is used during tagging it's automatically converted from the boolean to the string `True`.  I think the correct behaviour is that which you've added.  **But** I think we may need a note that Tags are automatically converted to Strings and you may need to explicitly search for the string "True".

jainnikhil30 commented:
@tremble 

Thanks for your review, it absolutely makes sense. I can/will work on adding those test cases you suggested. I may need some help with a few things, and I will reach out to you. I will update this PR.

jainnikhil30 commented:
@tremble 

I have made the changes you suggested. Apologies for those weird git commits its showing up. I am not sure what did I mess up while switching my fork from master to main. Also the shippable failure seems unrelated to to my changes ? The changes I have made:

- removed the ec2_ami_info target from integration tests and added those tests to ec2_ami
- added unit tests for ansible_dict_to_boto3_filter_list in test_ec2
- added a integration test in ec2_ami for checking string filters

Please let me know your thoughts on this.

tremble commented on changelogs/fragments/28-ec2_ami_info_int_bool_filter.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - ec2_ami_info - allow integer and bool values for filtering images (https://github.com/ansible/ansible/issues/43570).
```
```suggestion
  - ec2_ami_info - allow integer and bool values for filtering images (https://github.com/ansible/ansible/issues/43570).
  - ansible_dict_to_boto3_filter_list - convert integers and bools to strings before using them in filters
```

jainnikhil30 commented:
@tremble 

Thanks for that git tip !

tremble commented:
Failed test is a CI flake, the test didn't even start up.  Hopefully someone from the 'core'/'content' team will be able to retrigger the failed test.

jainnikhil30 commented:
@jillr @s-hertel 

Could you please take a look at this PR for your review ?

tremble commented:
Hi @jainnikhil30 - sorry we added some extra tests so there's a conflict, would you please rebase again

tremble commented:
If possible please would you add the string `ci_complete` to the final commit message, we *should* have automatically run most of the integration tests, but it didn't.  See also #110

jainnikhil30 commented:
@tremble 

Damn ! I think I messed up the commit messages again ! and it didnt run the ci_complete I think :(
Apologies for that. Do you want me to add another commit (adding a extra space or something else ) just for the ci_complete ?

tremble commented:
I minor whitespace change with ci_complete would be helpful if possible please.  There's the possibility that this would have an unexpected effect.  As long as the tests pass I think we're good to merge.

jainnikhil30 commented:
@tremble 

Did that work ?

tremble commented:
@jainnikhil30 the shippable test cases are the most important ones:

Full suite of test cases passed: https://app.shippable.com/github/ansible-collections/amazon.aws/runs/569/summary/console

You'll see that the AWS 'group' tests all took 8+ minutes (group 2 takes about 8-10 minutes, group 1 and 3 take closer to 12-15 minutes) which indicates that they all ran.

@jillr since you and @s-hertel already +1ed this I'm going to squash-merge and commit.

jillr commented:
For the future you can also do `git commit --allow-empty` and add a `ci_complete` message so you don't need to make minor changes. Thanks folks!
